File: src/main/utils/paths.ts
import { app } from 'electron'
import path from 'path'
import fs from 'fs'
import os from 'os'

/** 判断是否在开发模式 */
const isDev = !app.isPackaged

/**
 * 获取当前平台的 Kubo 二进制路径
 */
export function getKuboPath(): string {
  const platform = os.platform() // 'darwin' | 'win32' | 'linux'
  const arch = os.arch() // 'arm64' | 'x64' | 'ia32'

  // 构建文件名
  let filename: string
  if (platform === 'darwin') {
    filename = arch === 'arm64' ? 'ipfs-darwin-arm64' : 'ipfs-darwin-amd64'
  } else if (platform === 'win32') {
    filename = arch === 'arm64' ? 'ipfs-windows-arm64.exe' : 'ipfs-windows-amd64.exe'
  } else {
    // Linux 支持（可选）
    filename = arch === 'arm64' ? 'ipfs-linux-arm64' : 'ipfs-linux-amd64'
  }

  // 构建完整路径
  if (isDev) {
    return path.join(process.cwd(), 'resources', 'bin', filename)
  } else {
    return path.join(process.resourcesPath, 'bin', filename)
  }
}

/**
 * 验证 Kubo 二进制是否存在且可执行
 */
export function validateKuboPath(): { valid: boolean; path: string; error?: string } {
  const kuboPath = getKuboPath()
  const platform = os.platform()

  if (!fs.existsSync(kuboPath)) {
    return {
      valid: false,
      path: kuboPath,
      error: `Kubo binary not found at: ${kuboPath}`
    }
  }

  // macOS/Linux: 检查可执行权限
  if (platform !== 'win32') {
    try {
      fs.accessSync(kuboPath, fs.constants.X_OK)
    } catch {
      // 尝试修复权限
      fs.chmodSync(kuboPath, 0o755)
    }
  }

  return { valid: true, path: kuboPath }
}

/** IPFS repo 路径 */
export function getIPFSRepoPath(): string {
  const appData = app.getPath('appData')
  // macOS: ~/Library/Application Support/Planet/ipfs
  // Windows: %APPDATA%\Planet\ipfs
  const repoPath = path.join(appData, 'Planet', 'ipfs')
  fs.mkdirSync(repoPath, { recursive: true })
  return repoPath
}

/** 数据根目录 */
export function getDataPath(): string {
  const appData = app.getPath('appData')
  const dataPath = path.join(appData, 'Planet')
  fs.mkdirSync(dataPath, { recursive: true })
  return dataPath
}

/** 文档目录 */
export function getDocumentsPath(): string {
  const docs = app.getPath('documents')
  const planetDocs = path.join(docs, 'Planet')
  fs.mkdirSync(planetDocs, { recursive: true })
  return planetDocs
}

/** 临时目录 */
export function getTempPath(): string {
  const temp = app.getPath('temp')
  const planetTemp = path.join(temp, 'Planet')
  fs.mkdirSync(planetTemp, { recursive: true })
  return planetTemp
}

File: src-tauri/src/helpers/paths.rs
use std::path::PathBuf;
use tauri::{AppHandle, Manager};

/// 获取 Kubo 可执行文件路径（跨平台）
pub fn get_kubo_path(app: &AppHandle) -> PathBuf {
    // 在开发模式下，使用 target/debug/resources
    // 在生产模式下，使用 app.path().resource_dir()
    let resource_dir = if cfg!(debug_assertions) {
        // 开发模式：target/debug/resources
        std::env::current_exe()
            .ok()
            .and_then(|exe| exe.parent().map(|p| p.to_path_buf()))
            .map(|exe_dir| exe_dir.join("resources"))
            .unwrap_or_else(|| {
                // 如果上面的方法失败，尝试从当前工作目录推断
                std::env::current_dir()
                    .unwrap_or_default()
                    .join("target/debug/resources")
            })
    } else {
        // 生产模式：使用 Tauri 的资源目录
        app.path()
            .resource_dir()
            .expect("Failed to get resource dir")
    };

    #[cfg(target_os = "windows")]
    {
        #[cfg(target_arch = "x86_64")]
        {
            resource_dir.join("bin").join("kubo-windows-amd64.exe")
        }
        #[cfg(not(target_arch = "x86_64"))]
        {
            panic!("Unsupported Windows architecture")
        }
    }

    #[cfg(target_os = "macos")]
    {
        #[cfg(target_arch = "aarch64")]
        {
            resource_dir.join("bin").join("kubo-darwin-arm64")
        }
        #[cfg(target_arch = "x86_64")]
        {
            resource_dir.join("bin").join("kubo-darwin-amd64")
        }
        #[cfg(not(any(target_arch = "aarch64", target_arch = "x86_64")))]
        {
            panic!("Unsupported macOS architecture")
        }
    }

    #[cfg(target_os = "linux")]
    {
        resource_dir.join("bin").join("kubo-linux-amd64")
    }

    #[cfg(not(any(target_os = "windows", target_os = "macos", target_os = "linux")))]
    {
        panic!("Unsupported operating system")
    }
}

/// IPFS repo 路径
pub fn get_ipfs_repo_path(app: &AppHandle) -> PathBuf {
    let app_data = app
        .path()
        .app_data_dir()
        .expect("Failed to get app data dir");
    let repo = app_data.join("ipfs");
    std::fs::create_dir_all(&repo).ok();
    repo
}

/// 数据存储路径（Planet 数据）
pub fn get_data_path(app: &AppHandle) -> PathBuf {
    let app_data = app
        .path()
        .app_data_dir()
        .expect("Failed to get app data dir");
    let planet = app_data.join("Planet");
    std::fs::create_dir_all(&planet).ok();
    planet
}

/// 文档路径
pub fn get_documents_path(app: &AppHandle) -> PathBuf {
    app.path()
        .document_dir()
        .expect("Failed to get document dir")
}

/// 临时文件路径
pub fn get_temp_path(app: &AppHandle) -> PathBuf {
    let temp = app.path().temp_dir().expect("Failed to get temp dir");
    let planet_temp = temp.join("planet-desktop");
    std::fs::create_dir_all(&planet_temp).ok();
    planet_temp
}